src/keypair.rs

```diff
@@ -174,69 +169,131 @@ impl Deref for Key {
 
 // Don't make it implement SizedBytes so that it's not constructible outside of this module.
 impl Key {
-    fn to_arr(&self) -> GenericArray<u8, KeyLen> {
+    fn to_arr<L: ArrayLength<u8>>(&self) -> GenericArray<u8, L> {
         GenericArray::clone_from_slice(&self.0[..])
     }
 
     #[allow(clippy::unnecessary_wraps)]
-    fn from_arr(key_bytes: &GenericArray<u8, KeyLen>) -> Result<Self, TryFromSizedBytesError> {
+    fn from_arr<L: ArrayLength<u8>>(key_bytes: &GenericArray<u8, L>) -> Result<Self, TryFromSizedBytesError> {
         Ok(Key(key_bytes.to_vec()))
     }
 }
 
 /// Wrapper around a Key to enforce that it's a private one.
-#[derive(Debug, PartialEq, Eq, Clone, Hash, Zeroize)]
 #[cfg_attr(feature = "serialize", derive(serde::Deserialize, serde::Serialize))]
-// Ensure Key material is zeroed after use.
-#[zeroize(drop)]
 #[repr(transparent)]
-pub struct PrivateKey(Key);
+pub struct PrivateKey<G> {
+    key: Key,
+    _g: PhantomData<G>,
+}
+
+impl_clone_for!(
+    struct PrivateKey<G>,
+    [key, _g],
+);
+impl_debug_eq_hash_for!(
+    struct PrivateKey<G>,
+    [key, _g],
+);
 
-impl Deref for PrivateKey {
+// This can't be derived because of the use of a phantom parameter
+impl<G> Zeroize for PrivateKey<G> {
+    fn zeroize(&mut self) {
+        self.key.zeroize();
+    }
+}
+
+impl<G> Drop for PrivateKey<G> {
+    fn drop(&mut self) {
+        self.zeroize();
+    }
+}
+
+impl<G> Deref for PrivateKey<G> {
     type Target = Key;
 
     fn deref(&self) -> &Self::Target {
-        &self.0
+        &self.key
     }
 }
 
-impl SizedBytes for PrivateKey {
-    type Len = KeyLen;
+impl<G> PrivateKey<G> {
+    fn new(key: Key) -> Self {
+        Self {
+            key,
+            _g: PhantomData,
+        }
+    }
+}
+
+impl<G: Group> SizedBytes for PrivateKey<G> {
+    type Len = G::ScalarLen;
```

Comment:
I'm not entirely sure if this is correct: the `PrivateKey` has `Group::ScalarLen` size and the `PublicKey` has `Group::ElemLen` size.

Reply:
Yes!